opte-api/src/ip.rs

```diff
@@ -294,6 +332,17 @@ impl fmt::Display for IpAddr {
     }
 }
 
+impl FromStr for IpAddr {
+    type Err = String;
+    fn from_str(val: &str) -> result::Result<Self, Self::Err> {
+        if let Ok(ipv4) = val.parse::<Ipv4Addr>() {
+            Ok(ipv4.into())
+        } else {
+            val.parse::<Ipv6Addr>().map(IpAddr::Ip6)
```

Comment:
Thanks, that's a good point about the error. I can make it something version-agnostic.

But I'm not sure I agree with the part that we always know which version we're parsing from the command line. In this PR, I changed the `opteadm add-router-entry-ipv4` command to just `add-router-entry`. That accepts either an IPv4 or IPv6 address or CIDR block, which is exactly why I added this parsing code. It seemed better to simplify the CLI and the underlying `ioctl(2)` interface to accept an `IpAddr` rather than have nearly identical versions.

That said, I get the point -- it's exactly why we carry the IP version "up" through types like `Ipv4Cfg` and `Ipv6Cfg`, so that we can be sure at compile-time that the actual versions are the same for all data inside them. By _not_ doing that here, I had to add the runtime checks in the `valid_router_dest_target_pair` function, which checks that the destination and target are of the same protocol version. If you'd prefer, I can go the other way: keep IPv4 and IPv6 versions of the router APIs, and remove the runtime checks.

Reply:
Yea, and I hadn't even thought of the fact that we are pushing off dest+target validation to runtime. It seems like it could be nice to eliminate this runtime check from the router entry API. I haven't thought this through, but could we just give AddRouterEntryReq the Ip4/Ip6 enum treatment, and then the command line can either have two commands or one command with an option to indicate type.

That said, this is all nit picky and could be done in a follow up if we agree it would be cleaner this way. I don't want to thrash this PR too much. So I'll leave this up to you.

Reply:
I think I'm going to defer this. Making the `AddRouterEntryReq` into a V4 and V6 version means we also need to make the `RouterTarget` into the same. That's kinda lame, because the `Drop` and `InternetGateway` variants would be duplicated. I think the runtime check isn't great either, but it's not obvious how to make that much better without a bunch of other code.

Reply:
Okay, that's fine. I definitely don't want to hold this up.

Reply:
Awesome, thanks. I'll make an issue referring to this code once I merge. Running the last test to verify the ARP layer changes in a guest now.

Reply:
Ok, I've checked that guests behave the same way now, and that the right traffic is dropped (Inbound ARP other than the Proxy ARP requests, outbound other than for the gateway).